omniwitness/http.go

```diff
@@ -0,0 +1,192 @@
+		klog.V(1).Infof("Too many HTTP requests, pushing back.")
+		w.WriteHeader(http.StatusTooManyRequests)
+		return
+	}
+
+	oldSize, proof, cp, err := parseBody(r.Body)
+	if err != nil {
+		klog.V(1).Infof("invalid body: %v", err)
+		w.WriteHeader(http.StatusBadRequest)
+		counterHTTPIncomingResponse.Inc("unknown", strconv.Itoa(http.StatusBadRequest))
+		return
+	}
+	s := strings.SplitN(string(cp), "\n", 2)
+	if len(s) != 2 {
+		klog.V(1).Infof("invalid cp: %v", cp)
+		w.WriteHeader(http.StatusBadRequest)
+		counterHTTPIncomingResponse.Inc("unknown", strconv.Itoa(http.StatusBadRequest))
+		return
+	}
+
+	logID := log.ID(s[0])
+	logCfg, ok := a.logs[logID]
+	if !ok {
+		klog.V(1).Infof("unknown log: %v", logID)
+		w.WriteHeader(http.StatusNotFound)
+		counterHTTPIncomingResponse.Inc("unknown", strconv.Itoa(http.StatusNotFound))
+		return
+	}
+
+	sc, body, contentType, err := a.handleUpdate(r.Context(), logCfg.Origin, oldSize, cp, proof)
+	if err != nil {
+		klog.Errorf("handleUpdate: %v", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		counterHTTPIncomingResponse.Inc(logCfg.Origin, strconv.Itoa(http.StatusInternalServerError))
+		return
+	}
+
+	if contentType != "" {
+		w.Header().Add("Content-Type", contentType)
+	}
+	w.WriteHeader(sc)
+	if len(body) > 0 {
+		if _, err := w.Write(body); err != nil {
+			klog.Infof("Failed to write response body: %v", err)
+		}
+	}
+	counterHTTPIncomingResponse.Inc(logCfg.Origin, strconv.Itoa(sc))
+}
+
+// handleUpdate submits the provided checkpoint to the witness and interprets any errors which may result.
+//
+// Returns an appropriate HTTP status code, response body, and Content Type representing the outcome.
+func (a *httpHandler) handleUpdate(ctx context.Context, origin string, oldSize uint64, newCP []byte, proof [][]byte) (int, []byte, string, error) {
+	sigs, trustedSize, updateErr := a.update(ctx, oldSize, newCP, proof)
+	// Finally, handle any "soft" error from the update:
+	if updateErr != nil {
+		switch updateErr {
+		case witness.ErrCheckpointStale:
+			return http.StatusConflict, []byte(fmt.Sprintf("%d\n", trustedSize)), "text/x.tlog.size", nil
+		case witness.ErrUnknownLog:
+			return http.StatusNotFound, nil, "", nil
+		case witness.ErrNoValidSignature:
+			return http.StatusForbidden, nil, "", nil
+		case witness.ErrOldSizeInvalid:
+			return http.StatusBadRequest, nil, "", nil
+		case witness.ErrInvalidProof:
+			return http.StatusUnprocessableEntity, nil, "", nil
+		case witness.ErrRootMismatch:
+			return http.StatusConflict, nil, "", nil
+		default:
+			return http.StatusInternalServerError, nil, "", updateErr
+		}
+	}
+
+	return http.StatusOK, sigs, "", nil
+}
+
+// parseBody reads the incoming request and parses into constituent parts.
+//
+// The request body MUST be a sequence of
+// - a previous size line,
+// - zero or more consistency proof lines,
+// - and an empty line,
+// - followed by a [checkpoint][].
+func parseBody(r io.Reader) (uint64, [][]byte, []byte, error) {
```

Comment:
Could this be generally useful to have in https://github.com/transparency-dev/formats for those implementing witnesses, as it is specified by https://github.com/C2SP/C2SP/blob/main/tlog-witness.md?

(I was personally searching whether there was a package to unmarshal the "witness checkpoint update" body when implementing a simple "/add-checkpoint" endpoint for testing, and couldn't find anything - ended up stumbling upon this PR 😄 )